drivers/kscan/kscan_gpio.c

```diff
@@ -0,0 +1,495 @@
+
+		gpio_pin_set(row, row_cfg->pin, 1);
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			struct device *col = data->cols[c];
+			const struct kscan_gpio_item_config *col_cfg
+				= &cfg->cols[c];
+
+			read_state[r][c] =
+				gpio_pin_get(col, col_cfg->pin) > 0;
+		}
+
+		gpio_pin_set(row, row_cfg->pin, 0);
+	}
+
+	// Set all our outputs as active again, then re-enable interrupts,
+	// so we can trigger interrupts again for future press/release
+	kscan_gpio_set_row_state(dev, 1);
+	kscan_gpio_enable_interrupts(dev);
+
+	for (int r = 0; r < MATRIX_ROWS; r++) {
+		for (int c = 0; c < MATRIX_COLS; c++) {
+			bool pressed = read_state[r][c];
+
+			if (pressed != data->matrix_state[r][c]) {
+				LOG_DBG("Sending event at %d,%d state %s",
+					r, c, (pressed ? "on" : "off"));
+				data->matrix_state[r][c] = pressed;
+				data->callback(dev, r, c, pressed);
+			}
+		}
+	}
+
+	return 0;
+}
+
+#else
+
+static struct kscan_gpio_irq_callback irq_callbacks[MATRIX_ROWS];
+
+static int kscan_gpio_enable_interrupts(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+	const struct kscan_gpio_config *cfg = dev->config_info;
+
+	return kscan_gpio_config_interrupts(data->rows, cfg->rows, MATRIX_ROWS,
+			GPIO_INT_DEBOUNCE | GPIO_INT_EDGE_BOTH);
+}
+
+static int kscan_gpio_disable_interrupts(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+	const struct kscan_gpio_config *cfg = dev->config_info;
+
+	return kscan_gpio_config_interrupts(data->rows, cfg->rows, MATRIX_ROWS,
+			GPIO_INT_DISABLE);
+}
+
+static void kscan_gpio_set_column_state(struct device *dev, int value)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+	const struct kscan_gpio_config *cfg = dev->config_info;
+
+	for (int c = 0; c < MATRIX_COLS; c++) {
+		struct device *col = data->cols[c];
+		const struct kscan_gpio_item_config *col_cfg = &cfg->cols[c];
+
+		gpio_pin_set(col, col_cfg->pin, value);
+	}
+}
+
+static int kscan_gpio_read(struct device *dev)
+{
+	struct kscan_gpio_data *data = dev->driver_data;
+	const struct kscan_gpio_config *cfg = dev->config_info;
+
+	static bool read_state[MATRIX_ROWS][MATRIX_COLS];
+
+	LOG_DBG("Scanning the matrix for updated state");
+
+	// Disable our interrupts temporarily while we scan, to avoid
+	// re-entry while we iterate columns and set them active one by one
+	// to get pressed state for each matrix cell.
+	kscan_gpio_disable_interrupts(dev);
+	kscan_gpio_set_column_state(dev, 0);
+
+	for (int c = 0; c < MATRIX_COLS; c++) {
+		struct device *col = data->cols[c];
+		const struct kscan_gpio_item_config *col_cfg = &cfg->cols[c];
+
+		gpio_pin_set(col, col_cfg->pin, 1);
+
+		for (int r = 0; r < MATRIX_ROWS; r++) {
+			struct device *row = data->rows[r];
+			const struct kscan_gpio_item_config *row_cfg
+				= &cfg->rows[r];
+
+			read_state[r][c] =
+				gpio_pin_get(row, row_cfg->pin) > 0;
+		}
+
+		gpio_pin_set(col, col_cfg->pin, 0);
+	}
+
+	// Set all our outputs as active again, then re-enable interrupts,
```

Comment:
Change all your comments to /* */

Reply:
@fractalclone Done. (And rebased on latest `master` while at it to be safe).